### ISSUE TITLE
TypedDict doesn't work with overridden keys

### ISSUE THREAD

**Issue by chris-leach** [microsoft/pyright#7574]
**Describe the bug**
I'm trying to construct an instance of one `TypedDict` subclass from an instance of another subclass with many fields in common, using a dict comprehension with unpacking and overwriting the fields that are different, as in the commented line below.

It seems as though `make_test_2` contains all the necessary information for a static checker to verify that all fields are the correct type, since all fields with incompatible types are overwritten.

**Code or Screenshots**
```python
from typing import TypedDict


class Test1(TypedDict):
    field_1: str
    field_2: str


class Test2(TypedDict):
    field_1: str
    field_2: int


def make_test_1() -> Test1:
    return {
        "field_1": "abc",
        "field_2": "123",
    }


def make_test_2() -> Test2:
    test_1 = make_test_1()
    return {
        **test_1,  # Expression of type "dict[str, str | int]" cannot be assigned to return type "Test2"
        "field_2": int(test_1["field_2"]),
    }
```

**VS Code extension or command-line**
Pylance VS Code extension 2024.3.2
pyright 1.1.356

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug. 

I think it's a reasonable enhancement request, so I'm going to consider it in that light.

As you point out, the expression is writing an invalid type to a variable of type `Test2`, but this invalid type is shortly later overwritten with a valid type. It's similar to this code:

```python
def make_test_2() -> Test2:
    test_1 = make_test_1()

    temp: Test2 = {**test_1}
    temp["field_2"] = 1
    return temp
```

I think you'd agree that the above code should result in a type violation because `{**test_1}` is not type compatible with `Test2`. Even though you're overwriting the offending key in the next statement, it's still a type violation.

I'll need to think about this a bit more to determine whether there are any typing holes that could result from the proposed enhancement.

**Comment by chris-leach** [comment]
Thanks for the quick response!

I would definitely expect the second line of your function to incur an error, since the variable has an inline type hint. However, I would expect the following example to work:

```python
def make_test_2() -> Test2:
    test_1 = make_test_1()

    temp = {**test_1}  # not compliant with Test2, but no error since temp is not annotated
    temp["field_2"] = 1  # temp has changed (structural) type, and is now compliant with Test2
    return temp  # no error
```

**Comment by erictraut** [comment]
Your example above definitely won't type check. The expression `{*test_1}` by default is evaluated using normal dictionary inference rules, which means that its type is inferred as `dict[str, str]`. The next line should therefore result in a type error because assigning an `int` value to a `dict[str, str]` is illegal. Also, `dict[str, str]` is not type compatible with `Test2`, so the return statement should return an error here. This behavior is dictated by the Python typing spec, and pyright is compliant with the spec.

**Comment by chris-leach** [comment]
That’s fair enough - if the spec says variables can’t change type (even if the only type constraints are structural) then there’s not much to be done.

If there’s no such rule against the original example then I’ll continue to hold out hope for the improvement!

**Comment by erictraut** [comment]
This enhancement request will be included in the next release. 

https://github.com/microsoft/pyright/commit/c29473ab0f89e938972ff20fe00c71105ca59695

**Comment by chris-leach** [comment]
That's great, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
